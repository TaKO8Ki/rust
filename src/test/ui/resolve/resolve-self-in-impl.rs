#![feature(associated_type_defaults)]

struct S<T = u8>(T);
trait Tr<T = u8> {
    type A = ();
}

impl Tr<Self> for S {} // OK
impl<T: Tr<Self>> Tr<T> for S {} // OK
impl Tr for S where Self: Copy {} // OK
impl Tr for S where S<Self>: Copy {} // OK
impl Tr for S where Self::A: Copy {} // OK

impl Tr for Self {} //~ ERROR cycle detected
//~^ ERROR `Self` is only available in impls, traits, and type definitions
impl Tr for S<Self> {} //~ ERROR cycle detected
//~^ ERROR `Self` is only available in impls, traits, and type definitions
impl Self {} //~ ERROR cycle detected
//~^ ERROR `Self` is only available in impls, traits, and type definitions
impl S<Self> {} //~ ERROR cycle detected
//~^ ERROR `Self` is only available in impls, traits, and type definitions
impl Tr<Self::A> for S {} //~ ERROR cycle detected

fn main() {}
