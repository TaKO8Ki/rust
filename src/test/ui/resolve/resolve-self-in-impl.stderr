error: `Self` is only available in impls, traits, and type definitions
  --> $DIR/resolve-self-in-impl.rs:14:13
   |
LL | impl Tr for Self {}
   |             ^^^^

error: `Self` is only available in impls, traits, and type definitions
  --> $DIR/resolve-self-in-impl.rs:16:15
   |
LL | impl Tr for S<Self> {}
   |               ^^^^

error: `Self` is only available in impls, traits, and type definitions
  --> $DIR/resolve-self-in-impl.rs:18:6
   |
LL | impl Self {}
   |      ^^^^

error: `Self` is only available in impls, traits, and type definitions
  --> $DIR/resolve-self-in-impl.rs:20:8
   |
LL | impl S<Self> {}
   |        ^^^^

error[E0391]: cycle detected when computing type of `<impl at $DIR/resolve-self-in-impl.rs:14:1: 14:20>`
  --> $DIR/resolve-self-in-impl.rs:14:13
   |
LL | impl Tr for Self {}
   |             ^^^^
   |
   = note: ...which immediately requires computing type of `<impl at $DIR/resolve-self-in-impl.rs:14:1: 14:20>` again
note: cycle used when collecting item types in top-level module
  --> $DIR/resolve-self-in-impl.rs:1:1
   |
LL | / #![feature(associated_type_defaults)]
LL | |
LL | | struct S<T = u8>(T);
LL | | trait Tr<T = u8> {
...  |
LL | |
LL | | fn main() {}
   | |____________^

error[E0391]: cycle detected when computing type of `<impl at $DIR/resolve-self-in-impl.rs:16:1: 16:23>`
  --> $DIR/resolve-self-in-impl.rs:16:15
   |
LL | impl Tr for S<Self> {}
   |               ^^^^
   |
   = note: ...which immediately requires computing type of `<impl at $DIR/resolve-self-in-impl.rs:16:1: 16:23>` again
note: cycle used when collecting item types in top-level module
  --> $DIR/resolve-self-in-impl.rs:1:1
   |
LL | / #![feature(associated_type_defaults)]
LL | |
LL | | struct S<T = u8>(T);
LL | | trait Tr<T = u8> {
...  |
LL | |
LL | | fn main() {}
   | |____________^

error[E0391]: cycle detected when computing type of `<impl at $DIR/resolve-self-in-impl.rs:18:1: 18:13>`
  --> $DIR/resolve-self-in-impl.rs:18:6
   |
LL | impl Self {}
   |      ^^^^
   |
   = note: ...which immediately requires computing type of `<impl at $DIR/resolve-self-in-impl.rs:18:1: 18:13>` again
note: cycle used when collecting item types in top-level module
  --> $DIR/resolve-self-in-impl.rs:1:1
   |
LL | / #![feature(associated_type_defaults)]
LL | |
LL | | struct S<T = u8>(T);
LL | | trait Tr<T = u8> {
...  |
LL | |
LL | | fn main() {}
   | |____________^

error[E0391]: cycle detected when computing type of `<impl at $DIR/resolve-self-in-impl.rs:20:1: 20:16>`
  --> $DIR/resolve-self-in-impl.rs:20:8
   |
LL | impl S<Self> {}
   |        ^^^^
   |
   = note: ...which immediately requires computing type of `<impl at $DIR/resolve-self-in-impl.rs:20:1: 20:16>` again
note: cycle used when collecting item types in top-level module
  --> $DIR/resolve-self-in-impl.rs:1:1
   |
LL | / #![feature(associated_type_defaults)]
LL | |
LL | | struct S<T = u8>(T);
LL | | trait Tr<T = u8> {
...  |
LL | |
LL | | fn main() {}
   | |____________^

error[E0391]: cycle detected when computing trait implemented by `<impl at $DIR/resolve-self-in-impl.rs:22:1: 22:26>`
  --> $DIR/resolve-self-in-impl.rs:22:1
   |
LL | impl Tr<Self::A> for S {}
   | ^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: ...which immediately requires computing trait implemented by `<impl at $DIR/resolve-self-in-impl.rs:22:1: 22:26>` again
note: cycle used when collecting item types in top-level module
  --> $DIR/resolve-self-in-impl.rs:1:1
   |
LL | / #![feature(associated_type_defaults)]
LL | |
LL | | struct S<T = u8>(T);
LL | | trait Tr<T = u8> {
...  |
LL | |
LL | | fn main() {}
   | |____________^

error: aborting due to 9 previous errors

For more information about this error, try `rustc --explain E0391`.
